spmd/tensor/parallel/api.py

```diff
@@ -14,6 +14,176 @@
 from spmd.tensor.parallel import TensorParallelMultiheadAttention
 
 
+def _shard_input_1d(
+    tensor: Union[torch.Tensor, DTensor],
+    device_mesh: Optional[DeviceMesh] = None,
+    dim: int = 0,
+) -> DTensor:
+    if device_mesh is None:
+        if isinstance(DTensor, tensor):
+            device_mesh = tensor.device_mesh
+
+    assert device_mesh is not None, "_replicate_input_1d: device mesh is not 1D"
+
+    sharding = [Shard(dim)]
+    if isinstance(torch.Tensor, tensor):
+        return DTensor.from_local(tensor, device_mesh, sharding)
+    elif isinstance(DTensor, tensor):
+        return tensor.redistribute(device_mesh, sharding)
+    else:
+        raise RuntimeError(
+            "Input to _shard_input_1d must be either torch.Tensor or DTensor!"
+        )
+
+
+def MakeInputShard1D(dim: int) -> functools.partial[DTensor]:
+    # This function generates input handler for 1-D mesh device only
+    return functools.partial(_shard_input_1d, dim=dim)
+
+
+def _replicate_input_1d(
+    tensor: Union[torch.Tensor, DTensor],
+    device_mesh: Optional[DeviceMesh] = None,
+) -> DTensor:
+    if device_mesh is None:
+        if isinstance(DTensor, tensor):
+            device_mesh = tensor.device_mesh
+
+    assert device_mesh is not None, "_replicate_input_1d: device mesh is not 1D"
+
+    replicate = [Replicate()]
+    if isinstance(torch.Tensor, tensor):
+        return DTensor.from_local(tensor, device_mesh, replicate)
+    elif isinstance(DTensor, tensor):
+        return tensor.redistribute(device_mesh, replicate)
+    else:
+        raise RuntimeError(
+            "Input to _replicate_input_1d must be either torch.Tensor or DTensor!"
+        )
+
+
+def MakeInputReplicated() -> functools.partial[DTensor]:
+    # This function generates input handler for 1-D mesh device only
+    return functools.partial(_replicate_input_1d)
+
+
+def _replicate_output_1d(tensor: DTensor) -> torch.Tensor:
+    assert (
+        isinstance(tensor, DTensor) and tensor.device_mesh.ndim == 1
+    ), "_replicate_output_1d: device mesh is not 1D"
+    return replicate_output(tensor)
+
+
+def MakeOutputReplicated() -> functools.partial[torch.Tensor]:
+    # This function generates input handler for 1-D mesh device only
+    return functools.partial(_replicate_output_1d)
+
+
+@dataclass
+class ParallelStyle(object):
+    """
+    The parallel style user wants the module or submodule to be sharded.
+    Users can extend this class to build their own parallel style with customized input/output preparations.
+    """
+
+    _prepare_input: Callable[
+        [Union[torch.Tensor, DTensor], Optional[DeviceMesh]], DTensor
+    ]
+    _prepare_output: Callable[[DTensor], Union[Tensor, DTensor]]
```

Comment:
For output, we might also need a DeviceMesh, let's make it Optional. Also, since we need to pass in the dimension, can we also updated the signature here? I have updated the design already.